api/stocks.go

```diff
@@ -113,11 +210,199 @@ func (s *Server) fetchStockData(ctx context.Context, symbol string) (*TimeSeries
 		return nil, fmt.Errorf("error while reading response body: %w", err)
 	}
 
-	var data TimeSeriesDaily
+	var data T
 	err = json.Unmarshal(body, &data)
 	if err != nil {
 		return nil, fmt.Errorf("error unmarshaling JSON: %w", err)
 	}
 
 	return &data, nil
 }
+
+func calculateDebtEquityRatio(_ context.Context, balanceSheet *BalanceSheetMetadata) (float64, error) {
+	if len(balanceSheet.AnnualReports) == 0 {
+		return 0, errors.New("no available data")
+	}
+
+	recentReport := balanceSheet.AnnualReports[0]
+
+	totalLiabilities, err := strconv.ParseFloat(recentReport.TotalLiabilities, 64)
+	if err != nil {
+		return 0, err
+	}
+
+	totalEquity, err := strconv.ParseFloat(recentReport.TotalShareholderEquity, 64)
+	if err != nil {
+		return 0, err
+	}
+
+	if totalEquity == 0 {
+		return 0, errors.New("cannot division by zero")
+	}
+
+	ratio := totalLiabilities / totalEquity
+	return ratio, nil
+}
+
+func (h *StockHandler) updateStockMetrics(ctx context.Context, symbol string) ([]model.StockMetric, error) {
+	stock, err := h.store.FindStockBySymbol(ctx, symbol)
+	if err != nil {
+		if errors.Is(err, store.ErrNotFound) {
+			return nil, errors.New("stock not found")
+		}
+		return nil, fmt.Errorf("error while fetching stock by symbol: %w", err)
+	}
+
+	const maxNumMetric = 50
+	const offset = 0
+	metrics, err := h.store.ListMetrics(ctx, maxNumMetric, offset)
+	if err != nil {
+		return nil, err
+	}
+
+	metricMap := buildMetricMap(metrics)
+	overview, balanceSheet, fetchErr := h.combineStockData(ctx, symbol)
+	if fetchErr != nil {
+		return nil, fetchErr
+	}
+
+	var updatedStockMetrics []model.StockMetric
+	addMetric := h.metricAdder(ctx, stock, &updatedStockMetrics)
+
+	if balanceSheet != nil {
+		processBalanceSheetMetrics(ctx, balanceSheet, metricMap, addMetric, symbol)
+	}
+
+	if overview != nil {
+		processOverviewMetrics(ctx, overview, metricMap, addMetric)
+	}
+
+	return updatedStockMetrics, nil
+}
+
+func (h *StockHandler) combineStockData(
+	ctx context.Context,
+	symbol string,
+) (*OverviewMetadata, *BalanceSheetMetadata, error) {
+	type fetchResult struct {
+		overview     *OverviewMetadata
+		balanceSheet *BalanceSheetMetadata
+		err          error
+	}
+
+	const numCh = 2
+	resultCh := make(chan fetchResult, numCh)
+	var wg sync.WaitGroup
+	wg.Add(numCh)
+
+	go func() {
+		defer wg.Done()
+		overview, err := h.fetchStockOverview(ctx, symbol)
+		resultCh <- fetchResult{overview: overview, err: err}
+	}()
+
+	go func() {
+		defer wg.Done()
+		balanceSheet, err := h.fetchBalanceSheet(ctx, symbol)
+		resultCh <- fetchResult{balanceSheet: balanceSheet, err: err}
+	}()
+
+	go func() {
+		wg.Wait()
+		close(resultCh)
+	}()
+
+	var overview *OverviewMetadata
+	var balanceSheet *BalanceSheetMetadata
+	var fetchErrors []error
+
+	for res := range resultCh {
+		if res.err != nil {
+			fetchErrors = append(fetchErrors, res.err)
```

Comment:
```suggestion
	var errs error

	for res := range resultCh {
		if res.err != nil {
			errs = errors.Join(errs, res.err)
```

Go's standard `errors` package does that for you 😎